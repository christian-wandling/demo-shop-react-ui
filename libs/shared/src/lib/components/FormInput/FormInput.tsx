import { Controller, FieldError, useFormContext } from 'react-hook-form';
import { FormInputProps } from '../../models/formInputProps';
import { FormError } from '../FormError/FormError';

export const FormInput = ({
  name,
  label,
  placeholder,
  type = 'text',
  required = false,
  disabled = false,
  autoComplete,
  className,
  'data-testid': dataTestId,
  defaultValue = '',
}: FormInputProps) => {
  const {
    control,
    formState: { errors },
  } = useFormContext();
  const errorMessage = errors[name]?.message as string | undefined;

  return (
    <div className={className}>
      <label
        htmlFor={name}
        className={`block text-sm font-medium leading-6 ${disabled ? 'text-gray-500' : 'text-gray-900'}`}>
        {label} {required && <span className="required">*</span>}
      </label>
      <Controller
        name={name}
        control={control}
        defaultValue={defaultValue}
        render={({ field }) => (
          <input
            id={name}
            type={type}
            placeholder={placeholder}
            disabled={disabled}
            autoComplete={autoComplete}
            className={`block w-full rounded-md border-0 py-1.5  shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ${
              disabled ? 'text-gray-500 bg-gray-200' : 'text-gray-900'
            } ${errorMessage ? 'error' : ''} ${className || ''}`}
            data-testid={dataTestId}
            {...field}
          />
        )}
      />
      {errorMessage && <div className="error-message">{errorMessage}</div>}
      <FormError error={errors[name] as FieldError} />
    </div>
  );
};
