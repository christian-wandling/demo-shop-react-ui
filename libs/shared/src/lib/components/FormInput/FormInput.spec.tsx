import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { FormInput } from './FormInput';
import { useFormContext } from 'react-hook-form';

vi.mock('react-hook-form', () => ({
  useFormContext: vi.fn(),
  Controller: ({ render, defaultValue, name }: any) => {
    const field = {
      name,
      value: defaultValue,
      onChange: vi.fn(),
      onBlur: vi.fn(),
      ref: vi.fn(),
    };
    return render({ field });
  },
}));

vi.mock('../FormError/FormError', () => ({
  FormError: ({ error }: any) => (error ? <div data-testid="form-error">Mocked FormError</div> : null),
}));

describe('FormInput component', () => {
  const mockControl = { field: vi.fn() };
  const defaultProps = {
    name: 'testField',
    label: 'Test Field',
    placeholder: 'Enter test data',
  };

  beforeEach(() => {
    (useFormContext as any).mockReturnValue({
      control: mockControl,
      formState: { errors: {} },
    });
  });

  it('renders input field with label correctly', () => {
    render(<FormInput {...defaultProps} />);

    expect(screen.getByLabelText('Test Field')).toBeTruthy();
    expect(screen.getByPlaceholderText('Enter test data')).toBeTruthy();
    expect(screen.getByRole('textbox').id).toBe('testField');
  });

  it('matches the snapshot', () => {
    const { baseElement } = render(<FormInput {...defaultProps} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('displays required asterisk when required prop is true', () => {
    render(<FormInput {...defaultProps} required={true} />);

    const requiredIndicator = screen.getByText('*');

    expect(requiredIndicator).toBeTruthy();
  });

  it('applies disabled styling when disabled prop is true', () => {
    render(<FormInput {...defaultProps} disabled={true} />);

    const input: HTMLInputElement = screen.getByRole('textbox');

    expect(input.disabled).toBe(true);
    expect(input.classList).toContain('text-gray-500');
    expect(input.classList).toContain('bg-gray-200');
  });

  it('sets input type correctly', () => {
    render(<FormInput {...defaultProps} type="password" />);

    const input: HTMLInputElement = screen.getByLabelText('Test Field');
    expect(input.type).toBe('password');
  });

  it('displays error message when provided', () => {
    (useFormContext as any).mockReturnValue({
      control: mockControl,
      formState: {
        errors: {
          testField: {
            message: 'This field has an error',
            type: 'required',
          },
        },
      },
    });

    render(<FormInput {...defaultProps} />);

    expect(screen.getByText('This field has an error')).toBeTruthy();
    expect(screen.getByTestId('form-error')).toBeTruthy();
    expect(screen.getByRole('textbox').classList).toContain('error');
  });

  it('applies custom className when provided', () => {
    render(<FormInput {...defaultProps} className="custom-class" />);

    const inputContainer = screen.getByLabelText('Test Field').closest('div');
    expect(inputContainer).toBeTruthy();
    expect(inputContainer?.classList).toContain('custom-class');
  });

  it('sets default value correctly', () => {
    render(<FormInput {...defaultProps} defaultValue="Default Text" />);

    const input: HTMLInputElement = screen.getByRole('textbox');
    expect(input.value).toBe('Default Text');
  });

  it('sets autocomplete attribute when provided', () => {
    render(<FormInput {...defaultProps} autoComplete="email" />);

    const input: HTMLInputElement = screen.getByRole('textbox');
    expect(input.autocomplete).toBe('email');
  });
});
