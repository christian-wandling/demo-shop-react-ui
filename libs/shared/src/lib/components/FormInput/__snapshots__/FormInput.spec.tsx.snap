// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormInput component > matches the snapshot 1`] = `
<body>
  <div>
    <div>
      <label
        class="block text-sm font-medium leading-6 text-gray-900"
        for="testField"
      >
        Test Field
         
      </label>
      <input
        class="
              block w-full rounded-md border-0 py-1.5  shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6
              text-gray-900
               
            "
        id="testField"
        name="testField"
        placeholder="Enter test data"
        type="text"
        value=""
      />
    </div>
  </div>
</body>
`;
